npe2/_plugin_manager.py

```diff
@@ -52,105 +54,220 @@ def addi(self, begin: int, end: int, data: T) -> None:
+    ) -> Iterator[ReaderContribution]:
+        if not path:
+            return
+
+        if isinstance(path, list):
+            if len({Path(i).suffix for i in path}) > 1:
+                raise ValueError(
+                    "All paths in the stack list must have the same extension."
+                )
+            path = path[0]
+
+        if os.path.isdir(path):
+            yield from self._readers[""]
+        else:
+            seen: Set[str] = set()
+            for ext, readers in self._readers.items():
+                if ext and fnmatch(str(path), ext):
+                    for r in readers:
+                        if r.command not in seen:
+                            seen.add(r.command)
+                            yield r
+
+    def iter_compatible_writers(
+        self, layer_types: Sequence[str]
+    ) -> Iterator[WriterContribution]:
+        """Attempt to match writers that consume all layers."""
+
+        if not layer_types:
+            return
+
+        # First count how many of each distinct type are requested. We'll use
+        # this to get candidate writers compatible with the requested count.
+        counts = Counter(layer_types)
+
+        def _get_candidates(lt: LayerType) -> Set[WriterContribution]:
+            return {v.data for v in self._writers[lt][counts[lt]] or []}
+
+        types = iter(LayerType)
+        candidates = _get_candidates(next(types))
+        for lt in types:
+            if candidates:
+                candidates &= _get_candidates(lt)
+            else:
+                break
+
+        def _writer_key(writer: WriterContribution) -> Tuple[bool, int, int, List[str]]:
+            # 1. writers with no file extensions (like directory writers) go last
+            no_ext = len(writer.filename_extensions) == 0
+
+            # 2. more "specific" writers first
+            nbounds = sum(not c.is_zero() for c in writer.layer_type_constraints())
+
+            # 3. then sort by the number of listed extensions
+            #    (empty set of extensions goes last)
+            ext_len = len(writer.filename_extensions)
+
+            # 4. finally group related extensions together
+            exts = writer.filename_extensions
+            return (no_ext, nbounds, ext_len, exts)
+
+        yield from sorted(
+            candidates,
+            key=_writer_key,
+        )
+
+
+class PluginManagerEvents(SignalGroup):
+    plugins_registered = Signal(
+        set,
+        description="Emitted with a set of PluginManifest instances "
+        "whenever new plugins are registered. 'Registered' means that a "
+        "manifest has been provided or discovered.",
+    )
+    activation_changed = Signal(
+        set,
+        set,
+        description="Emitted with two arguments: a set of plugin "
+        "names that were activated, and a set of names that were "
+        "deactivated. 'Activated' means the plugin has been *imported*, it's "
+        "`on_activate` function was called.",
```

Comment:
```suggestion
        "deactivated. 'Activated' means the plugin has been *imported*, its "
        "`on_activate` function was called.",
```

I think... I'm unsure. I doubt now.

Reply:
yes,  you're right :joy:  this is one of those things that my brain knows and my fingers almost always do wrong :)

Reply:
["I heard you're idea's and their definately good"](https://xkcd.com/1238/)

Honestly it might be easier for non-native speakers.